### DISCUSSION TITLE
How to use Strimzi for local development?

### DISCUSSION THREAD

**Discussion by iirekm** [strimzi/strimzi-kafka-operator#8124] [Q&A]
While setuping Kafka with docker-compose works out of the box, with Strimzi I get errors like:
```
kafkacat -b my-cluster-kafka-external-bootstrap.kafka:9094 -t x -P 
%3|1676735228.667|FAIL|rdkafka#producer-1| [thrd:my-cluster-kafka-external-bootstrap.kafka:9094/bootstrap]: my-cluster-kafka-external-bootstrap.kafka:9094/bootstrap: Failed to resolve 'my-cluster-kafka-external-bootstrap.kafka:9094': Name or service not known (after 13ms in state CONNECT)
```
I tried both with hardcoding DNS in /etc/hosts, with getting it via Telepresence, with k3s kubernetes, with kind kubernetes nothing works. Exactly the same method of setting up any other service over k8s (MySQL, PostgreSQL, Spring Boot, RabbitMQ) is as easy as pie, but not for Strimzi :-(

**Comment by scholzj** [comment]
You will have to share more about your setup to get any help. How did you configure the Kafka cluster, what does the environment you run it in look like, what did you do to get it working, etc. You can also check many of the discussions here on the same topic. Kafka has its own discovery mechanism which allows it to scale. So it is not as easy as with the other services.

**Comment by iirekm** [comment]
The examples you provided on website - all of them use `kubectl` to access Kafka. How to access Kafka without kubectl? IntelliJ, PyCharm, VS Code, local Spring Boot, local Python, kafkacat, Kafka IDE,  and other developer tools DON'T RUN inside k8s, they are local processes that have no idea about k8s services. I tried using `kubectl port-forward` or `telepresence connect` - none of them works in Strimzi's case. It seems something is blocking access to Kafka from outside of k8s cluster.
The documentation is inadequate. Please provide instructions how to DEVELOP with Strimzi, not only how to put it into production. With `docker-compose` getting development-ready Kafka is easy (just use `localhost:9092`). Why it's so problematic with Strizi?

OS: Ubuntu 22
Kubernetes: kind, k3s (on both the same problems)

**Reply by scholzj** [reply]
You didn't really answer the questions I asked TBH. So it is a bit hard to help. 
* Telepresence does not handle DNS of headless services properly. So it does not work with the `type: internal` listeners. It might work with the `type: cluster-ip` style listener - I did not try it lately, but that might work.
* Port forwarding does not work with Kafka because as I explained earlier, Kafka has its own discovery mechanisms. If you wanna learn more about that, you can for example check out [this blog post series](https://strimzi.io/blog/2019/04/17/accessing-kafka-part-1/) which is about accessing a Kafka cluster from outside the Kubernetes cluster (which is what you most likely have to use in your case as well regardless whether you are in production or not).
* I do not use `kind` or `k3s` locally and do not run on Ubuntu. So I do not know how that environment works in detail TBH. You will likely need to configure an external listener to make it work. With Minikube using a VM driver, the node port listener normally works. Yu can try Kubernetes Nginx Ingress with TLS Passthrough as well. But each environment has many options on how to run them and all of it depends on your operating system it is hard to provide a definitive guide.

> The documentation is inadequate. Please provide instructions on how to DEVELOP with Strimzi, not only how to put it into production. With docker-compose getting development-ready Kafka is easy (just use localhost:9092). Why it's so problematic with Strizi?

TBH, the easiest way to run Kafk for development is just locally from the Kafka binaries ZIP file 🤷. When you add Kubernetes into the mix, you get quite a complex situation where different networks mix together, separate DNS addressing, etc. So you need to properly configure the Kafka advertised hosts according to the environment. And that is not easy because every operating system, every local Kubernetes cluster tool etc. behaves differently and what works on my laptop will not work for you etc. As explained in the blog posts I linked above -> The Kafka protocol and its discovery mechanisms allow it to scale, but also make some things hard. That is also why documenting it is so hard because there is no single way to do things that would work everywhere.

----

If you really have a use case where you need to run it on Kubernetes, it can be usually configured the right way with some effort. But you have to stop shouting that you want someone to write step-by-step documentation for you and instead start sharing back what you did or didn't do. And after few iterations, we might get there. If you don't have strong reasons why to use Kubernetes for this, then TBH running Kafka locally or using Docker compose might be simply easier.
